pkg/querysharding/analyzer.go

```diff
@@ -5,41 +5,92 @@ package querysharding
 
 import (
 	"fmt"
+	"os"
 
+	"github.com/go-kit/log"
+	lru "github.com/hashicorp/golang-lru"
 	"github.com/prometheus/prometheus/promql/parser"
 )
 
 // QueryAnalyzer is an analyzer which determines
 // whether a PromQL Query is shardable and using which labels.
 type QueryAnalyzer struct{}
 
+var LRU *lru.Cache
+var logger log.Logger
+
 var nonShardableFuncs = []string{
 	"label_join",
 	"label_replace",
 }
 
 // NewQueryAnalyzer creates a new QueryAnalyzer.
 func NewQueryAnalyzer() *QueryAnalyzer {
+	inIT() // Initializing cache
 	return &QueryAnalyzer{}
 }
 
+func inIT() {
```

Comment:
We can just move the init function to `NewQueryAnalyzer`, no need a separate function.
And we should return the error.